tasks/RedundantRBSSelectorTask.cpp

```diff
@@ -0,0 +1,164 @@
+/* Generated from orogen/lib/orogen/templates/tasks/Task.cpp */
+
+#include "RedundantRBSSelectorTask.hpp"
+
+using namespace transforms;
+
+bool rbsIsValid(base::samples::RigidBodyState const& rbs);
+
+RedundantRBSSelectorTask::RedundantRBSSelectorTask(std::string const& name)
+    : RedundantRBSSelectorTaskBase(name)
+{
+}
+
+RedundantRBSSelectorTask::~RedundantRBSSelectorTask()
+{
+}
+
+/// The following lines are template definitions for the various state machine
+// hooks defined by Orocos::RTT. See RedundantRBSSelectorTask.hpp for more detailed
+// documentation about them.
+
+bool RedundantRBSSelectorTask::configureHook()
+{
+    if (!RedundantRBSSelectorTaskBase::configureHook())
+        return false;
+    m_source_timeout = _source_timeout.get();
+    m_main_source_hysteresis = _source_timeout.get();
+    return true;
+}
+bool RedundantRBSSelectorTask::startHook()
+{
+    if (!RedundantRBSSelectorTaskBase::startHook())
+        return false;
+
+    m_internal_state = InternalState();
+    m_internal_state.main_source_deadline = m_internal_state.secondary_source_deadline =
+        base::Time::now() + m_source_timeout;
+    m_internal_state.hysteresis_deadline = base::Time::now();
+
+    return true;
+}
+
+RedundantRBSSelectorTask::States updateState(
+    RedundantRBSSelectorTask::States nominal_state,
+    RedundantRBSSelectorTask::InternalState internal)
+{
+    // for readability's sake
+    using States = RedundantRBSSelectorTask::States;
+    base::Time now = base::Time::now();
+
+    switch (nominal_state) {
+        case States::RUNNING:
+            if (now > internal.main_source_deadline) {
+                return updateState(States::INVALID_MAIN_SOURCE, internal);
+            }
+
+            if (now > internal.secondary_source_deadline) {
+                return updateState(States::REDUNDANCY_OFF, internal);
+            }
+        case States::HYSTERESIS_WAIT:
+            if (now > internal.hysteresis_deadline ||
+                now > internal.secondary_source_deadline) {
+                return updateState(States::RUNNING, internal);
+            }
```

Comment:
The point of this state is to verify that the main source remains valid for a long time before switching back. You need to test this

```
if (now > internal.main_source_deadline) {
    return updateState(States::INVALID_MAIN_SOURCE, internal);
}
// then the check that you've written
```